bad-bits/lib/store.js

```diff
@@ -7,38 +7,42 @@
  * @param {string} etag - ETag for the current denylist
  */
 export async function updateBadBitsDatabase(env, currentHashes, etag) {
-  try {
-    const now = new Date()
-    const insertBadBitStmt = env.DB.prepare(
-      `
+  const startedAt = Date.now()
+  const now = new Date().toISOString()
+  const insertBadBitStmt = env.DB.prepare(
+    `
       INSERT INTO bad_bits (hash, last_modified_at) VALUES (?, ?)
       ON CONFLICT(hash) DO UPDATE SET last_modified_at = excluded.last_modified_at
       `,
-    )
-
-    const statements = [
-      ...Array.from(currentHashes).map((hash) =>
-        insertBadBitStmt.bind(hash, now.toISOString()),
-      ),
+  )
 
-      env.DB.prepare('DELETE FROM bad_bits WHERE last_modified_at < ?').bind(
-        now.toISOString(),
-      ),
-    ]
+  let updated = 0
+  const remainingHashes = Array.from(currentHashes)
+  while (remainingHashes.length > 0) {
+    // pop first N hashes from remainingHashes
+    const batchHashes = remainingHashes.splice(0, 10_000)
+    updated += batchHashes.length
+    await env.DB.batch(
+      batchHashes.map((hash) => insertBadBitStmt.bind(hash, now)),
+    )
+    console.log(
+      `Inserted/updated ${updated} bad bits in ${Date.now() - startedAt}ms`,
+    )
```

Comment:
Should we keep this log for diagnostics, or is it just noise that's better to remove?

Reply:
to me fine either way, happy to follow up if I later think it's noisy